plugin/manager/manager.go

```diff
@@ -0,0 +1,269 @@
+	p, err := mgr.newPlugin(ctx, name)
+	if err != nil {
+		return nil, pluginErr(name, err)
+	}
+	if p.Err != nil {
+		return nil, pluginErr(name, withErr(p.Err, ErrNotCompliant))
+	}
+	if cmd == plugin.CommandGetMetadata {
+		return &p.Metadata, nil
+	}
+	var data []byte
+	if req != nil {
+		data, err = json.Marshal(req)
+		if err != nil {
+			return nil, pluginErr(name, fmt.Errorf("failed to marshal request object: %w", err))
+		}
+	}
+	resp, err := run(ctx, mgr.cmder, p.Path, cmd, data)
+	if err != nil {
+		return nil, pluginErr(name, err)
+	}
+	return resp, nil
+}
+
+// newPlugin determines if the given candidate is valid and returns a Plugin.
+func (mgr *Manager) newPlugin(ctx context.Context, name string) (*Plugin, error) {
+	ok := isCandidate(mgr.fsys, name)
+	if !ok {
+		return nil, ErrNotFound
+	}
+
+	p := &Plugin{Path: binPath(mgr.fsys, name)}
+	out, err := run(ctx, mgr.cmder, p.Path, plugin.CommandGetMetadata, nil)
+	if err != nil {
+		p.Err = fmt.Errorf("failed to fetch metadata: %w", err)
+		return p, nil
+	}
+	p.Metadata = *out.(*plugin.Metadata)
+	if p.Name != name {
+		p.Err = fmt.Errorf("executable name must be %q instead of %q", addExeSuffix(plugin.Prefix+p.Name), filepath.Base(p.Path))
+	} else if err := p.Metadata.Validate(); err != nil {
+		p.Err = fmt.Errorf("invalid metadata: %w", err)
+	}
+	return p, nil
+}
+
+// run executes the command and decodes the response.
+func run(ctx context.Context, cmder commander, pluginPath string, cmd plugin.Command, req []byte) (interface{}, error) {
+	out, ok, err := cmder.Output(ctx, pluginPath, string(cmd), req)
+	if err != nil {
+		return nil, fmt.Errorf("failed running the plugin: %w", err)
+	}
+	if !ok {
+		var re plugin.RequestError
+		err = json.Unmarshal(out, &re)
+		if err != nil {
+			return nil, withErr(plugin.RequestError{Code: plugin.ErrorCodeGeneric, Err: err}, ErrNotCompliant)
+		}
+		return nil, re
+	}
+	var resp interface{}
+	switch cmd {
+	case plugin.CommandGetMetadata:
+		resp = new(plugin.Metadata)
+	case plugin.CommandGenerateSignature:
+		resp = new(plugin.GenerateSignatureResponse)
+	case plugin.CommandGenerateEnvelope:
+		resp = new(plugin.GenerateEnvelopeResponse)
+	default:
+		return nil, fmt.Errorf("unsupported command: %s", cmd)
+	}
+	err = json.Unmarshal(out, resp)
+	if err != nil {
+		err = fmt.Errorf("failed to decode json response: %w", err)
+		return nil, withErr(err, ErrNotCompliant)
+	}
+	return resp, nil
+}
+
+func pluginErr(name string, err error) error {
+	return fmt.Errorf("%s: %w", name, err)
+}
+
+// isCandidate checks if the named plugin is a valid candidate.
+func isCandidate(fsys fs.FS, name string) bool {
+	base := binName(name)
+	fi, err := fs.Stat(fsys, path.Join(name, base))
+	if err != nil {
+		// Ignore any file which we cannot Stat
+		// (e.g. due to permissions or anything else).
+		return false
+	}
+	if fi.Mode().Type() != 0 {
+		// Ignore non-regular files.
+		return false
+	}
```

Comment:
We should additionally check if the containing dir `~/.notation/plugins/{plugin-name}/notation-{plugin-name}` is also not a symlink, something that was missed in the spec.

Reply:
This is exactly what it is being checked here. Non-regular files include symlinks and directories.

Reply:
I'm bit confused here. Isn't this code only checking that the plugin executable (`notation-{plugin-name}`) itself is not a symlink, and not that additionally the enclosing directory is also not a symlink (`~/.notation/plugins/{plugin-name}` dir)?